File: configmasterhelper/src/main/java/com/spascoding/configmasterhelper/javahelper/ConfigMasterHelper.java
package com.spascoding.configmasterhelper.javahelper;

import android.content.ContentValues;
import android.content.Context;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.util.Log;

import org.json.JSONObject;

public class ConfigMasterHelper {

    private static final Uri CONFIG_URI =
            Uri.parse("content://com.spascoding.configmaster.data.provider.ConfigProvider");

    /**
     * Insert or update configuration into ConfigMaster
     */
    public static void insertConfig(Context context, String configName, String jsonData) {

        if (!isPackageInstalled(context, "com.spascoding.configmaster")) {
            return;
        }

        ContentValues values = new ContentValues();
        values.put("configName", configName);
        values.put("jsonData", jsonData);

        context.getContentResolver().insert(CONFIG_URI, values);
    }

    /**
     * Fetch configuration from ConfigMaster
     */
    public static String fetchConfig(Context context, String configName) {
        Cursor cursor = context.getContentResolver().query(
                CONFIG_URI,
                null,
                null,
                new String[]{configName},
                null
        );

        String jsonData = null;
        if (cursor != null) {
            try {
                if (cursor.moveToFirst()) {
                    jsonData = cursor.getString(cursor.getColumnIndexOrThrow("jsonData"));
                    Log.d("ConfigMasterHelper", "Fetched config = " + jsonData);
                }
            } finally {
                cursor.close();
            }
        }
        return jsonData;
    }

    /**
     * Fetch a single parameter value by key from configuration JSON
     */
    public static String fetchConfigParam(Context context, String configName, String paramKey) {
        String jsonData = fetchConfig(context, configName);
        if (jsonData == null || jsonData.isEmpty()) {
            return null;
        }

        try {
            JSONObject jsonObject = new JSONObject(jsonData);
            if (jsonObject.has(paramKey)) {
                return jsonObject.getString(paramKey);
            } else {
                return null;
            }
        } catch (Exception e) {
            Log.e("ConfigMasterHelper", "Error parsing JSON: " + e.getMessage());
            return null;
        }
    }

    private static boolean isPackageInstalled(Context context, String packageName) {
        PackageManager packageManager = context.getPackageManager();
        try {
            packageManager.getPackageInfo(packageName, 0);
            return true;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }
}

File: configmasterhelper/src/main/java/com/spascoding/configmasterhelper/ConfigMasterHelper.kt
package com.spascoding.configmasterhelper

import android.content.ContentValues
import android.content.Context
import android.content.pm.PackageManager
import android.net.Uri
import org.json.JSONObject

object ConfigMasterHelper {

    private val CONFIG_URI: Uri =
        Uri.parse("content://com.spascoding.configmaster.data.provider.ConfigProvider")

    // ===================== PROVIDER COMMUNICATION =====================

    fun insertConfig(context: Context, configName: String, jsonData: String) {
        if (!isPackageInstalled(context, "com.spascoding.configmaster")) return

        val values = ContentValues().apply {
            put("configName", configName)
            put("jsonData", jsonData)
        }
        context.contentResolver.insert(CONFIG_URI, values)
    }

    fun fetchConfig(context: Context, configName: String): String? {
        val cursor = context.contentResolver.query(
            CONFIG_URI,
            null,
            null,
            arrayOf(configName),
            null
        )

        var jsonData: String? = null
        cursor?.use {
            if (it.moveToFirst()) {
                jsonData = it.getString(it.getColumnIndexOrThrow("jsonData"))
            }
        }
        return jsonData
    }

    private fun isPackageInstalled(context: Context, packageName: String): Boolean {
        return try {
            context.packageManager.getPackageInfo(packageName, 0)
            true
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }
    }

    // ===================== CONFIG CACHE MANAGER =====================
    object Parameter {

        // configName -> (key -> value)
        private val cache = mutableMapOf<String, MutableMap<String, String>>()

        /**
         * MUST be called ONCE from Application or Activity
         */
        fun init(context: Context, vararg configNames: String) {
            configNames.forEach { configName ->
                preloadConfig(context, configName)
            }
        }

        private fun preloadConfig(context: Context, configName: String) {
            val json = fetchConfig(context, configName) ?: return

            val map = mutableMapOf<String, String>()
            val obj = JSONObject(json)

            obj.keys().forEach { key ->
                map[key] = obj.getString(key)
            }

            cache[configName] = map
        }

        /**
         * ✅ Access anywhere - NO CONTEXT REQUIRED
         */
        fun get(configName: String, key: String): String? {
            return cache[configName]?.get(key)
        }

        fun getAll(configName: String): Map<String, String> {
            return cache[configName] ?: emptyMap()
        }

        fun refresh(context: Context, configName: String) {
            preloadConfig(context, configName)
        }
    }
}